src/mpi/romio/adio/ad_gpfs/ad_gpfs_aggrs.c

```diff
@@ -709,54 +699,57 @@ void ADIOI_GPFS_Calc_others_req(ADIO_File fd, int count_my_req_procs,
             others_req[i].lens = NULL;
         }
     }
-    /* If no recv buffer was allocated in the loop above, make it NULL */
-    if (recvBuf == (void *) 0xFFFFFFFFFFFFFFFF)
-        recvBuf = NULL;
 
     /* Now send the calculated offsets and lengths to respective processes */
 
     /************************/
     /* Exchange the offsets */
     /************************/
 
-    /* Determine the lowest sendBuf */
     for (i = 0; i < nprocs; i++) {
-        if ((my_req[i].count) && ((uintptr_t) my_req[i].offsets <= (uintptr_t) sendBuf)) {
-            sendBuf = my_req[i].offsets;
-        }
-        /* my_req[i].offsets and my_req[i].lens have been malloc-ed together */
+        scounts[i] = count_my_req_per_proc[i] * 2;
+        rcounts[i] = count_others_req_per_proc[i] * 2;
     }
 
-    /* If no send buffer was found in the loop above, make it NULL */
-    if (sendBuf == (void *) 0xFFFFFFFFFFFFFFFF)
-        sendBuf = NULL;
-
-    /* Calculate the displacements from the sendBuf */
+    /*
+     * This code used to set sendBuf to the lowest my_req[i].offsets and
+     * then set sdisps[] to the relative offsets of the other my_req[].offsets
+     * values, but those are all separate malloc()s and could be anywhere in
+     * memory so those displacements won't necessarily fit in an integer.
+     *
+     * So I'm creating a new buffer, copying the data into it, then copying
+     * the data back out after the Alltoallv.
+     */
+    void *sbuf_copy_of_req_info;
+    void *rbuf_copy_of_req_info;
+    int scount_total = 0;
+    int rcount_total = 0;
     for (i = 0; i < nprocs; i++) {
-        /* Send these offsets and lengths to process i. */
-        scounts[i] = count_my_req_per_proc[i] * 2;
-        if (scounts[i] == 0)
-            sdispls[i] = 0;
-        else
-            sdispls[i] = (int)
-                (((uintptr_t) my_req[i].offsets -
-                  (uintptr_t) sendBuf) / (uintptr_t) sizeof(ADIO_Offset));
-
-        /* Receive these offsets and lengths from process i. */
-        rcounts[i] = count_others_req_per_proc[i] * 2;
-        if (rcounts[i] == 0)
-            rdispls[i] = 0;
-        else
-            rdispls[i] = (int)
-                (((uintptr_t) others_req[i].offsets -
-                  (uintptr_t) recvBuf) / (uintptr_t) sizeof(ADIO_Offset));
+        sdispls[i] = scount_total;
+        scount_total += scounts[i];
+        rdispls[i] = rcount_total;
+        rcount_total += rcounts[i];
+    }
+    sbuf_copy_of_req_info = (ADIO_Offset *) ADIOI_Malloc(scount_total * sizeof(ADIO_Offset));
+    rbuf_copy_of_req_info = (ADIO_Offset *) ADIOI_Malloc(rcount_total * sizeof(ADIO_Offset));
+    for (i = 0; i < nprocs; i++) {
+        memcpy(sbuf_copy_of_req_info + sdispls[i] * sizeof(ADIO_Offset),
+               my_req[i].offsets, scounts[i] * sizeof(ADIO_Offset));
```

Comment:
Should we concern about the extra memcpy overhead plus the extra memory footprint, considering some applications may do huge buffer operations.

There is MPI-4 API coming that adds `MPI_Alltoallv_c` to allow large count displs array. Maybe we can wait for that.